fastparquet/writer.py

```diff
@@ -747,15 +800,116 @@ def partition_on_columns(data, columns, root_path, partname, fmd, sep,
             fmd.row_groups.append(rg)
 
 
-def write_common_metadata(fn, fmd, open_with=default_openw):
-    """For hive-style parquet, write schema in special shared file"""
-    if isinstance(fn, str):
-        f = open_with(fn)
-    else:
-        f = fn
-    f.write(MARKER)
-    fmd.row_groups = []
-    foot_size = write_thrift(f, fmd)
-    f.write(struct.pack(b"<i", foot_size))
-    f.write(MARKER)
-    f.close()
+def write_common_metadata(fn, fmd, open_with=default_open,
+                          no_row_groups=True):
+    """
+    For hive-style parquet, write schema in special shared file
+
+    Parameters
+    ----------
+    fn: str
+        Filename to write to
+    fmd: thrift FileMetaData
+        Information to write
+    open_with: func
+        To use to create writable file as f(path, mode)
+    no_row_groups: bool (True)
+        Strip out row groups from metadata before writing - used for "common
+        metadata" files, containing only the schema.
+    """
+    with open_with(fn, 'wb') as f:
+        f.write(MARKER)
+        if no_row_groups:
+            rgs = fmd.row_groups
+            fmd.row_groups = []
+            foot_size = write_thrift(f, fmd)
+            fmd.row_groups = rgs
+        else:
+            foot_size = write_thrift(f, fmd)
+        f.write(struct.pack(b"<i", foot_size))
+        f.write(MARKER)
+
+
+def merge(file_list, verify_schema=True, open_with=default_open):
+    """
+    Create a logical data-set out of multiple parquet files.
+
+    The files referenced in file_list must either be in the same directory,
+    or at the same level within a structured directory, where the directories
+    give partitioning information. The schemas of the files should also be
+    consistent.
+
+    Parameters
+    ----------
+    file_list: list of paths
+    verify_schema: bool (True)
+        If True, will first check that all the schemas in the input files are
+        identical.
+    open_with: func
+        Used for opening a file for writing as f(path, mode)
+    """
+    sep = sep_from_open(open_with)
+    basepath, file_list = analyse_paths(file_list, sep)
+
+    pf = api.ParquetFile(sep.join([basepath, file_list[0]]),
+                         open_with=open_with)
+
+    if verify_schema:
+        # we don't open and store the ParquetFile instances, because the
+        # operation may be memory expensive, e.g., for s3fs back-end.
+        for fn in file_list[1:]:
+            pfi = api.ParquetFile(sep.join([basepath, fn]),
+                                  open_with=open_with)
+            if pfi.schema != pf.schema:
+                raise ValueError('Incompatible schemas')
+
+    fmd = pf.fmd  # we inherit "created by" field
+
+    for rg in pf.row_groups:
+        for chunk in rg.columns:
+            chunk.file_path = file_list[0]
```

Comment:
Can this logic be combined with the logic below?  I suspect that this is a trade off between performance and code complexity.